watson/frames.py

```diff
@@ -143,13 +143,28 @@ def new_frame(self, project, start, stop, tags=None, id=None,
     def dump(self):
         return tuple(frame.dump() for frame in self._rows)
 
-    def filter(self, projects=None, tags=None, span=None):
-        return (
-            frame for frame in self._rows
-            if (projects is None or frame.project in projects) and
-               (tags is None or any(tag in frame.tags for tag in tags)) and
-               (span is None or frame in span)
-        )
+    def filter(self, projects=None, tags=None, span=None,
+               include_partial_frames=False):
+        for frame in self._rows:
+            if not (projects is None or frame.project in projects):
+                continue
+            if not (tags is None or any(tag in frame.tags for tag in tags)):
+                continue
+            if span is None or frame in span:
+                yield frame
+                continue
+
+            # If requested, return the part the frame that is within span, for
+            # frames that are *partially* within span or reaching over span
+            if span is not None and include_partial_frames:
+                if frame.start < span.start and frame.stop > span.stop:
+                    yield frame._replace(start=span.start, stop=span.stop)
+                elif (frame.start < span.start and
+                      (span.start < frame.stop < span.stop)):
+                    yield frame._replace(start=span.start)
+                elif ((span.start < frame.start < span.stop) and
+                      frame.stop > span.stop):
+                    yield frame._replace(stop=span.stop)
```

Comment:
I think this would work as well (replacing my previous suggestion as well):
```python
if span is not None and if frame not in span:
    if include_partial_frames:
        if (span.start < frame.start < span.stop or span.start < frame.stop < span.stop):
            frame = frame._replace(start=max(frame.start, span.start), stop=min(frame.stop, span.stop))
    continue

yield frame
```

Reply:
@k4nar I think that if you don't yield the result of `frame._replace()` you're going to lose it, because of the `continue` after it. Also, I'm not sure if the `if` inside the first condition is what you intended (`...and if frame not...`).